### PR TITLE
Fix: Arguments Break when Updating

### PR DESCRIPTION
**Pull request by Yohoki**
Edit: The method is slightly changed now. But the outcome is still the same. (https://github.com/Ryujinx/Ryujinx/pull/3744#issuecomment-1271276850)

Original Post:
Ryujinx updates fail to load any arguments that have spaces in them.

Example, loading Ryujinx with command line arguments:
````
-r "D:/Custom Ryujinx Folder"
arg[0] = "-r"
arg[1] = "D:/Custom Ryujinx Folder"
````
Will restart after updates with:
````
-r D:/Custom Ryujinx Folder
arg[0] = "-r"
arg[1] = "D:/Custom"
arg[2] = "Ryujinx"
arg[3] = "Folder"
````

This now wraps all arguments in quotes, resulting in:
````
"-r" "D:/Custom Ryujinx Folder"
arg[0] = "-r"
arg[1] = "D:/Custom Ryujinx Folder"
````

The quotes around each of the args gets ignored when run (eg. "-r" still works fine).

edit:
I should note that this was tested in windows. I have not tested it under Ubuntu. It's a small edit, though. Should be easy to get it working for Ubuntu if this fails and I'd be glad to try figuring that out if it needs worked on.
Edit2:
~Wasn't able to quickly figure out how to fake an update to fully test before getting busy today. But I was able to confirm that the ryuArg string does show as described, and I was able to start Ryujinx(windows) successfully with the string formatted as such.~ Got around to faking the update routine. Got a "Would you like to update...", clicked yes, watched dialog happen and Ryujinx successfully restart with the desired effect. Logs below.

### CODE REVIEW

**Comment by Yohoki** [comment]
Was able to fake an update and run the update routine successfully. Ryujinx "updated", restarted, and loaded the correct folder, with spaces and all.

Console log shows 2 startups, both with correct folder, spaces included. There were some errors that cropped up, after commenting out a LOT of code to fake the update. They're unrelated to what I actually pushed and include my real name, so I've redacted them.
````
00:00:00.003 |I| Configuration LogValueChange: EnableFileLog set to: False
00:00:00.005 |I| Configuration LogValueChange: ResScale set to: 1
00:00:00.006 |I| Configuration LogValueChange: ResScaleCustom set to: 0.75
00:00:00.006 |I| Configuration LogValueChange: MaxAnisotropy set to: 2
00:00:00.007 |I| Configuration LogValueChange: AspectRatio set to: Fixed16x9
00:00:00.008 |I| Configuration LogValueChange: BackendThreading set to: On
00:00:00.009 |I| Configuration LogValueChange: GraphicsBackend set to: OpenGl
00:00:00.010 |I| Configuration LogValueChange: PreferredGpu set to:
00:00:00.018 |N| Application PrintSystemInfo: Ryujinx Version: 1.0.0-dirty
00:00:00.022 |N| Application Print: Operating System: Microsoft Windows 10.0.19044 (X64)
00:00:00.022 |N| Application Print: CPU: Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz ; 8 logical
00:00:00.023 |N| Application Print: RAM: Total 8144 MB ; Available 1198 MB
00:00:00.027 |N| Application PrintSystemInfo: Logs Enabled: Error
00:00:00.027 |N| Application PrintSystemInfo: Launch Mode: Custom Path D:\Ryujinx Data Folder
//Redacted some build related errors that included names.
//Updater resets Ryujinx here.
00:00:00.005 |I| Configuration LogValueChange: EnableFileLog set to: False
00:00:00.007 |I| Configuration LogValueChange: ResScale set to: 1
00:00:00.007 |I| Configuration LogValueChange: ResScaleCustom set to: 0.75
00:00:00.007 |I| Configuration LogValueChange: MaxAnisotropy set to: 2
00:00:00.008 |I| Configuration LogValueChange: AspectRatio set to: Fixed16x9
00:00:00.010 |I| Configuration LogValueChange: BackendThreading set to: On
00:00:00.011 |I| Configuration LogValueChange: GraphicsBackend set to: OpenGl
00:00:00.011 |I| Configuration LogValueChange: PreferredGpu set to:
00:00:00.021 |N| Application PrintSystemInfo: Ryujinx Version: 1.0.0-dirty
00:00:00.025 |N| Application Print: Operating System: Microsoft Windows 10.0.19044 (X64)
00:00:00.025 |N| Application Print: CPU: Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz ; 8 logical
00:00:00.026 |N| Application Print: RAM: Total 8144 MB ; Available 1182 MB
00:00:00.029 |N| Application PrintSystemInfo: Logs Enabled: Error
00:00:00.029 |N| Application PrintSystemInfo: Launch Mode: Custom Path D:\Ryujinx Data Folder
````

**Comment by mageven** [comment]
Why join the args at all? Does this overload not work?

https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.start?view=net-6.0#system-diagnostics-process-start(system-string-system-collections-generic-ienumerable((system-string)))

**Comment by Yohoki** [comment]
The code is already (and still) using Process.Start(filename, args). That works fine, Ryujinxs is restarted and the args are passed to the new instance.

The issue is that when the args are passed, the quotes don't come with them. So paths with whitespace are cut off and loaded as multiple strings, instead of a single one. Example, the arguments '-r "D:/Custom Ryujinx Folder" ' wil startup ryujinx correctly, but after updating will be passed as '-r D:/Custom Ryujinx Folder ' with no quotes, resulting in 4 args being parsed, instead of 2.
````
-r D:/Custom Ryujinx Folder
arg[0] = "-r"
arg[1] = "D:/Custom"
arg[2] = "Ryujinx"
arg[3] = "Folder"
````

**Comment by mageven** [comment]
> The issue is that when the args are passed, the quotes don't come with them. So paths with whitespace are cut off and loaded as multiple strings, instead of a single one. 

Yes. This happens because of the join. That's why I suggested not manually joining them. Let the framework handle the best way for each OS.

**Comment by Yohoki** [comment]
I can try that out real quick and see. The only reason I used Join is because that's how it was already done and I wanted to make minimal changes.

**Comment by Yohoki** [comment]
OK. Made that change.

ryuArgs is now a string[] and passes each of the variables as individual strings instead of a single joined one.

Still updates correctly in windows. Ubuntu untested.

**Comment by Yohoki** [comment]
Changed to draft, as #3707 is ready to be merged, and affects my changes. I'll redo the changes when that merge has gone through, since my edits are a lot simpler and don't add any new features.

File: Ryujinx.Ava/Modules/Updater/Updater.cs [hunk 1]
```diff
@@ -278,7 +278,7 @@ public static async void UpdateRyujinx(Window parent, string downloadUrl)
                 {
                     string ryuName = Path.GetFileName(Environment.ProcessPath);
                     string ryuExe = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, ryuName);
-                    string ryuArg = string.Join(" ", Environment.GetCommandLineArgs().Skip(1).ToArray());
+                    string[] ryuArg = Environment.GetCommandLineArgs().Skip(1).ToArray();
```
Comment (mageven) [inline Ryujinx.Ava/Modules/Updater/Updater.cs:281]: Cloning it is unnecessary. The function already takes an `IEnumerable`
```suggestion
                    var ryuArg = Environment.GetCommandLineArgs().Skip(1);
```
Or just skip the variable and use it directly. Same for the other.

**Comment by mageven** [comment]
Personally, a simple bugfix is likely merged faster over a refactor/enhancement as it's easier to review. Maintainers will ask for a rebase on the conflicting PR so you don't have to worry about such details.

File: Ryujinx.Ava/Modules/Updater/Updater.cs (hunk 1 above)
Comment (Yohoki) [reply Ryujinx.Ava/Modules/Updater/Updater.cs]: Fair enough. I've set both to var instead.

I'd rather leave the declaration, though. It is clear what it does and easy to understand.

**Comment by Yohoki** [comment]
I'm not familiar with all this github stuff, this is my first big project that I've worked on. Should I be continuously merging in the main every time it updates, and is there anything I need to do to get the required checks done?

**Comment by MutantAura** [comment]
Usually a rebase/master merge when asked by a group member or when there are now conflicts. Merging every time isn't really needed, especially if there is nothing conflicting.

2 reviews and approvals are required to merge so just sit tight until someone can take a look :)

**Comment by Yohoki** [comment]
kk. Just making sure I wasn't sitting still when everyone was waiting for me to do something. I'll stop touching all the things then. :)

**Review by marysaka — APPROVED** [review]

**Review by AcK77 — APPROVED** [review]

**Pull request merged**